### PR TITLE
Use proper error message when session write fails #20807

### PR DESCRIPTION
**Pull request by digilist**
This improves the error message that is thrown if a session write fails (see #20807)

As there was no way to get the actual handler, I introduced a method on the SessionHandlerProxy. I hope that's okay, otherwise please give me a hint on what to do.

| Q             | A
| ------------- | ---
| Branch?       | 2.8
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #20807
| License       | MIT

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Session/Storage/Proxy/SessionHandlerProxy.php [hunk 1]
```diff
@@ -35,6 +35,14 @@ public function __construct(\SessionHandlerInterface $handler)
         $this->saveHandlerName = $this->wrapper ? ini_get('session.save_handler') : 'user';
     }
 
+    /**
+     * @return \SessionHandlerInterface
+     */
+    public function getHandler(): \SessionHandlerInterface
```
Comment (stof) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Proxy/SessionHandlerProxy.php]: The return type is a no-go as Symfony supports PHP 5.x

File: src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php [hunk 2]
```diff
@@ -223,7 +223,29 @@ public function regenerate($destroy = false, $lifetime = null)
      */
     public function save()
     {
-        session_write_close();
+        try {
+            session_write_close();
+        } catch (\ErrorException $e) {
```
Comment (stof) [inline src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php]: this won't work in prod, as notices and warnings are *not* turned into ErrorException exceptions there.

Code must not be written by assuming that the error handler has this behavior

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by nicolas-grekas** [comment]
Status: needs work

File: src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php (hunk 2 above)
Comment (digilist) [inline src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php]: Yes, currently the session write fails silently in production.

I updated the code to check the last error message and throw a E_USER_WARNING with the corrected message. However, the warning is not logged by default.

Alternatively, we can throw an Excpetion if the write fails, as it is an error anyway. However, it does break BC.

**Comment by nicolas-grekas** [comment]
Instead of relying on some external error handler to trigger the exception, I'd suggest to register a temporary error handler that does the throw.

**Comment by digilist** [comment]
That is a good idea! The code is much cleaner now.

File: src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php [hunk 3]
```diff
@@ -223,7 +224,30 @@ public function regenerate($destroy = false, $lifetime = null)
...
+            restore_error_handler();
+        } catch (ContextErrorException $e) {
+            // The default PHP error message is not very helpful, as it does not give any information on the current save handler.
+            // Therefore, we catch this error and trigger a warning with a better error message
+            $handler = $this->getSaveHandler();
+            if ($handler instanceof SessionHandlerProxy) {
+                $handler = $handler->getHandler();
+            }
+
+            $message = sprintf(
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php]: this can be inlined in the `trigger_error()` call

**Comment by fabpot** [comment]
Thank you @digilist.

**Pull request closed without merging**